### PR TITLE
Improve: try to fit simple list/array elements on a single line

### PR DESCRIPTION
**Pull request by gpetiot**
Will fix issue #368.

### CODE REVIEW

**Comment by gpetiot** [comment]
I did not see anything awful with `make -C test test test_diff` but it looks like the code of infer is well behind the current state of ocamlformat so the diff is huge and it is easy to miss something.

**Comment by gpetiot** [comment]
It does not change anything, maybe I missed something. Anyway I added an option `--compact-expressions` for this feature (the doc will be added soon).

**Comment by jberdine** [comment]
The dirs under test/code are just normal git repos, so you could use normal git commands to see if they are up to date. An additional technique is to run the test with the merge base of your feature branch, then stage the changes with test_stage, then rerun the test with your feature branch and check the diff with test_diff.

File: src/Conf.ml [hunk 1]
```diff
@@ -376,6 +377,13 @@ module Formatting = struct
     C.choice ~names ~all ~doc ~section (fun conf x ->
         {conf with break_struct= Poly.(x = `Force)} )
 
+  let compact_expressions =
+    let default = false in
+    let doc = "Style for lists and arrays." in
```
Comment (jberdine) [inline src/Conf.ml]: This could be a bit more descriptive.

File: src/Conf.ml [hunk 2]
```diff
@@ -376,6 +377,13 @@ module Formatting = struct
...
+    let names = ["compact-expressions"] in
```
Comment (jberdine) [inline src/Conf.ml]: "expressions" is probably overly general, maybe "literals" instead?

File: src/Fmt_ast.ml [hunk 3]
```diff
@@ -90,6 +90,16 @@ let update_config c l =
...
   List.fold ~init:c l ~f:update_one
 
+let fmt_expressions c width to_exp exprs fmt fmt_expr =
+  let grps =
+    List.group exprs ~break:(fun x1 x2 ->
+        (not (is_simple c.conf width (to_exp x1)))
+        || not (is_simple c.conf width (to_exp x2)) )
+  in
+  let fmt_grp exprs = hovbox (-2) (list exprs "@,; " fmt_expr) in
+  if c.conf.compact_expressions then list grps fmt fmt_grp
```
Comment (jberdine) [inline src/Fmt_ast.ml]: Can the computation of grps be sunk into the then branch?

File: src/Conf.ml (hunk 2 above)
Comment (gpetiot) [reply src/Conf.ml]: I don't know if "literals" is suited. I was thinking of "compact-linear-data-structures", it designates lists+arrays and not much more, but it is much longer than "compact-lists-arrays" and less explicit.

File: src/Conf.ml [hunk 4]
```diff
@@ -376,6 +377,13 @@ module Formatting = struct
...
+    C.flag ~default ~names ~doc ~section (fun conf compact_expressions ->
+        {conf with compact_expressions} )
```
Comment (jberdine) [inline src/Conf.ml]: What about handling this like `break-struct`, where the representation is a bool but the option accepts named alternates? For this option, I think using the same choices as `break-infix` would work well. That would also make it easier to give better docs.

File: src/Conf.ml (hunk 2 above)
Comment (jberdine) [reply src/Conf.ml]: Yes, "literal" is wrong. Perhaps "break-collection-expressions"?

Also, we could imagine applying this setting to tuples as well.

**Comment by gpetiot** [comment]
I added a fix for default-to-true flags, maybe it would have been better to have it in another PR, but it's so small that I think it is not so bad to put it here.

File: src/Conf.ml [hunk 5]
```diff
@@ -221,18 +222,19 @@ end = struct
...
     in
     let doc = generated_flag_doc ~allow_inline ~doc ~section in
     let docs = section_name section in
-    let term = Arg.(value & flag & info names_for_cmdline ~doc ~docs) in
+    let term = Arg.(value & flag & info names ~doc ~docs) in
     let parse = Bool.of_string in
     let r = mk ~default term in
-    let cmdline_get () = if !r then Some (not invert_flag) else None in
+    let cmdline_get () = if !r then Some true else None in
+    let update c x = update c (if invert_flag then not x else x) in
```
[the remark this answers is not in the document]
Comment (gpetiot) [reply src/Conf.ml]: Mmh sorry, I thought it was intended to have `no-` options in attributes as well when I tried to write `[@@@ocamlformat "no-break-collection-expressions"]` in commit 8df73c5. I will revert it.

File: src/Fmt_ast.ml [hunk 6]
```diff
@@ -1554,9 +1566,11 @@ and fmt_expression c ?(box = true) ?epi ?eol ?parens ?ext
         ( wrap_fits_breaks "[|" "|]" (Cmts.fmt_within c.cmts pexp_loc)
         $ fmt_atrs )
   | Pexp_array e1N ->
+      let e1N = List.map e1N ~f:(sub_exp ~ctx) in
       hvbox 0
         ( wrap_fits_breaks "[|" "|]"
-            (list e1N "@;<0 1>; " (sub_exp ~ctx >> fmt_expression c))
+            (fmt_expressions c width Fn.id e1N "@;<0 1>; "
```
[the remark this answers is not in the document]
Comment (gpetiot) [reply src/Fmt_ast.ml]: I don't get the same output though, so I will keep it that way for the moment.

File: src/Fmt_ast.ml [hunk 7]
```diff
@@ -90,6 +90,18 @@ let update_config c l =
   in
   List.fold ~init:c l ~f:update_one
 
+let fmt_expressions c width to_exp to_xt exprs fmt fmt_expr =
```
[the remark this answers is not in the document]
Comment (gpetiot) [reply src/Fmt_ast.ml]: Well I don't see what you have in mind for this function, as the types are not the same for lists and for arrays, maybe is was not necessary to factorize it in the first place then?

[the remark this answers is not in the document]
Comment (jberdine) [reply src/Fmt_ast.ml]: Ok, I see. Maybe just replace to_exp and to_xt with one sub_exp function?

**Pull request merged**